bayesian_changepoint_detection_experiment/01_change_point_detection.py

```diff
@@ -0,0 +1,248 @@
+    ]
+)
+print(full_token_set_df.shape)
+full_token_set_df.head()
+
+# # Combine both information
+
+# Using an idea similar to SCAF \[[1](https://doi.org/10.1007/s00799-019-00271-6)\], I'm combining the global qst metric with the percent change in frequency.
+# SCAF uses percent change for both metrics; however, the caveat is that their method loses information for the first two timepoints.
+# In my case given that global_distance_qst is a metric that's bound between 0 and 1 frequency percent change can be used directly.
+# By combining these two terms I can use this metric as a means to estimate change and allow for bayesian changepoint detection to calculate the probability of a timepoint change.
+
+merged_frequency_df = (
+    full_token_set_df
+    >> ply_tdy.unite("year", "year_1", "year_2", sep="-")
+    >> ply.inner_join(all_word_pct_change_df, on=["tok", "year"])
+    >> ply.rename(year_pair="year")
+    >> ply.select(
+        "tok",
+        "original_global_distance",
+        "global_distance_qst",
+        "year_pair",
+        "frequency",
+        "freq_pct_change",
+    )
+)
+merged_frequency_df
+
+change_metric_df = (
+    merged_frequency_df
+    >> ply.group_by("tok")
+    >> ply.arrange("year_pair")
+    >> ply.define(change_metric="global_distance_qst + freq_pct_change")
+    >> ply.ungroup()
+)
+change_metric_df
+
+# # Change point detection
+
+# Use Semantic Change Analysis with Frequency (SCAF) to perform bayesian change point detection.
+
+if not Path("output/bayesian_changepoint_data.tsv", sep="\t").exists():
+    change_point_results = []
+    for tok, tok_series_df in tqdm.tqdm(change_metric_df.groupby("tok")):
+        change_metric = np.insert(tok_series_df >> ply.pull("change_metric"), 0, 0)
+        Q, P, Pcp = offcd.offline_changepoint_detection(
+            change_metric,
+            partial(offcd.const_prior, l=(len(change_metric) + 1)),
+            offcd.gaussian_obs_log_likelihood,
+            truncate=-40,
+        )
+        estimated_changepoint_probability = np.exp(Pcp).sum(0)
+
+        change_point_results.append(
+            pd.DataFrame.from_dict(
+                {
+                    "tok": [tok] * len(estimated_changepoint_probability),
+                    "changepoint_prob": estimated_changepoint_probability,
+                    "year_pair": (
+                        tok_series_df
+                        >> ply.select("year_pair")
+                        >> ply.pull("year_pair")
+                    ),
+                }
+            )
+        )
+
+if not Path("output/bayesian_changepoint_data.tsv", sep="\t").exists():
+    change_point_df = pd.concat(change_point_results)
+    change_point_df.to_csv(
+        "output/bayesian_changepoint_data.tsv", sep="\t", index=False
+    )
+else:
+    change_point_df = pd.read_csv("output/bayesian_changepoint_data.tsv", sep="\t")
+change_point_df.head()
+
+change_point_df >> ply.arrange("-changepoint_prob") >> ply.slice_rows(30)
```

Comment:
>seems like there's a lot of German words in here. It seems strange that "zu" would have a change point, since it's a preposition (to/towards). maybe this is driven by a change in number of german articles?

Yeah the issue here is that a lot of abstracts back in 200 have a german translation right along with the english version. Trying to filter those examples out could hurt the model more, so its a cofounding issue that I'm conscious of moving forward. 
Also, this listing is only for words that appear in all years. I have yet to explore words that originate in later years.